cli/internal/oauth/api/api.go

```diff
@@ -0,0 +1,156 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/docker/cli/cli/internal/oauth/util"
+)
+
+type OAuthAPI interface {
+	GetDeviceCode(ctx context.Context, audience string) (State, error)
+	WaitForDeviceToken(ctx context.Context, state State) (TokenResponse, error)
+	Refresh(ctx context.Context, token string) (TokenResponse, error)
+	LogoutURL() string
+}
+
+// API represents API interactions with Auth0.
+type API struct {
+	// BaseURL is the base used for each request to Auth0.
+	BaseURL string
+	// ClientID is the client ID for the application to auth with the tenant.
+	ClientID string
+	// Scopes are the scopes that are requested during the device auth flow.
+	Scopes []string
+	// Client is the client that is used for calls.
+	Client util.Client
+}
+
+// TokenResponse represents the response of the /oauth/token route.
+type TokenResponse struct {
+	AccessToken      string  `json:"access_token"`
+	IDToken          string  `json:"id_token"`
+	RefreshToken     string  `json:"refresh_token"`
+	Scope            string  `json:"scope"`
+	ExpiresIn        int     `json:"expires_in"`
+	TokenType        string  `json:"token_type"`
+	Error            *string `json:"error,omitempty"`
+	ErrorDescription string  `json:"error_description,omitempty"`
+}
+
+var ErrTimeout = errors.New("timed out waiting for device token")
+
+// GetDeviceCode returns device code authorization information from Auth0.
+func (a API) GetDeviceCode(ctx context.Context, audience string) (state State, err error) {
+	data := url.Values{
+		"client_id": {a.ClientID},
+		"audience":  {audience},
+		"scope":     {strings.Join(a.Scopes, " ")},
+	}
+
+	deviceCodeURL := a.BaseURL + "/oauth/device/code"
```

Comment:
**suggestion**: could we maybe do `url.JoinPath` instead?
```suggestion
	deviceCodeURL, err := url.JoinPath(a.BaseURL, "/oauth/device/code")
```

Reply:
Why? It's slower than string concatenation and for no benefit here.

Reply:
This isn't a blocker to me, but thought I'd call it out. Since `BaseURL` is exported, we don't _know_ if it is formatted correctly. 

My thinking around this is more to keep the URL from becoming malformed by accident (for whatever reason). I guess here we don't read `BaseURL` from a config file or Env file, but if we did we could end up with a malformed URL without noticing it.